### PR TITLE
Build PreservedLabelState when triggering evition in RB/CRB application controller

### PR DESCRIPTION
**Pull request by XiShanYongYe-Chang**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Build PreservedLabelState when triggering evition in RB/CRB application controller

**Which issue(s) this PR fixes**:
Part of #5788

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
karmada-controller-manager: build PreservedLabelState when triggering evition in RB/CRB application controller
```

### CODE REVIEW

**Review by RainbowMango — COMMENTED** [review]
/assign

File: pkg/apis/work/v1alpha2/binding_types.go [hunk 1]
```diff
@@ -300,6 +300,9 @@ type GracefulEvictionTask struct {
 	// Populated by the system. Read-only.
 	// +optional
 	CreationTimestamp *metav1.Time `json:"creationTimestamp,omitempty"`
+
+	// ClusterBeforeFailover records the clusters where running the application before failover.
+	ClusterBeforeFailover []string `json:"clusterBeforeFailover,omitempty"`
```
[the remark this answers is not in the document]
Comment (XiShanYongYe-Chang) [reply pkg/apis/work/v1alpha2/binding_types.go:305]: Thanks for your remind!

File: pkg/controllers/applicationfailover/common.go [hunk 2]
```diff
@@ -117,3 +125,106 @@ func distinguishUnhealthyClustersWithOthers(aggregatedStatusItems []workv1alpha2
...
+func buildTaskOptions(failoverBehavior *policyv1alpha1.ApplicationFailoverBehavior, aggregatedStatus []workv1alpha2.AggregatedStatusItem, cluster, producer string, clustersBeforeFailover []string) ([]workv1alpha2.Option, error) {
+	var taskOpts []workv1alpha2.Option
+	taskOpts = append(taskOpts, workv1alpha2.WithProducer(producer))
+	taskOpts = append(taskOpts, workv1alpha2.WithReason(workv1alpha2.EvictionReasonApplicationFailure))
+	taskOpts = append(taskOpts, workv1alpha2.WithPurgeMode(failoverBehavior.PurgeMode))
+
+	if failoverBehavior.StatePreservation != nil && len(failoverBehavior.StatePreservation.Rules) != 0 {
+		targetStatusItem, exist := findTargetStatusItemByCluster(aggregatedStatus, cluster)
+		if !exist || targetStatusItem.Status == nil || targetStatusItem.Status.Raw == nil {
+			return nil, fmt.Errorf("the binding status under Cluster(%s) has not been colloected yet ", cluster)
```
Comment (RainbowMango) [inline pkg/controllers/applicationfailover/common.go:198]:
```suggestion
			return nil, fmt.Errorf("the application status has not yet been collected from cluster: %s", cluster)
```
Fixed typo `colloected`.

File: pkg/controllers/applicationfailover/crb_application_failover_controller.go [hunk 3]
```diff
@@ -153,40 +150,15 @@ func (c *CRBApplicationFailoverController) syncBinding(ctx context.Context, bind
...
-				return err
-			}
-		case policyv1alpha1.Immediately:
-			binding.Spec.GracefulEvictCluster(cluster, workv1alpha2.NewTaskOptions(
-				workv1alpha2.WithPurgeMode(policyv1alpha1.Immediately),
-				workv1alpha2.WithProducer(CRBApplicationFailoverControllerName),
-				workv1alpha2.WithReason(workv1alpha2.EvictionReasonApplicationFailure)))
+		taskOpts, err := buildTaskOptions(binding.Spec.Failover.Application, binding.Status.AggregatedStatus, cluster, CRBApplicationFailoverControllerName, clustersBeforeFailover)
+		if err != nil {
+			klog.Errorf("failed to build TaskOptions for ResourceBinding(%s/%s) under Cluster(%s): %v", binding.Namespace, binding.Name, cluster, err)
```
Comment (RainbowMango) [inline pkg/controllers/applicationfailover/crb_application_failover_controller.go:157]:
```suggestion
			klog.Errorf("failed to build TaskOptions for ClusterResourceBinding(%s/%s) under Cluster(%s): %v", binding.Namespace, binding.Name, cluster, err)
```

File: pkg/controllers/applicationfailover/common.go (hunk 2 above)
Comment (XiShanYongYe-Chang) [reply pkg/controllers/applicationfailover/common.go]: Updated, thx~

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Pull request merged**